server/service/async/async.go

```diff
@@ -0,0 +1,281 @@
+			}
+		}
+
+		// just ignore if there is no valid host id
+		if hostID == 0 {
+			return nil, nil, nil
+		}
+
+		conn := pool.ConfigureDoer(pool.Get())
+		defer conn.Close()
+
+		for {
+			vals, err := redigo.Ints(conn.Do("ZPOPMIN", key, redisPopCount))
+			if err != nil {
+				return nil, nil, errors.Wrap(err, "redis ZPOPMIN")
+			}
+			items := len(vals) / 2 // each item has the label id and the score (-1=delete, +1=insert)
+			stats.Items += items
+
+			for i := 0; i < len(vals); i += 2 {
+				labelID := vals[i]
+
+				var score int
+				if i+1 < len(vals) { // just to be safe we received all pairs
+					score = vals[i+1]
+				}
+
+				switch score {
+				case 1:
+					inserts = append(inserts, [2]uint{uint(labelID), hostID})
+				case -1:
+					deletes = append(deletes, [2]uint{uint(labelID), hostID})
+				}
+			}
+			if items < redisPopCount {
+				return inserts, deletes, nil
+			}
+		}
+	}
+
+	// Based on those pages, the best approach appears to be INSERT with multiple
+	// rows in the VALUES section (short of doing LOAD FILE, which we can't):
+	// https://www.databasejournal.com/features/mysql/optimize-mysql-inserts-using-batch-processing.html
+	// https://dev.mysql.com/doc/refman/5.7/en/insert-optimization.html
+	// https://dev.mysql.com/doc/refman/5.7/en/optimizing-innodb-bulk-data-loading.html
+	//
+	// Given that there are no UNIQUE constraints in label_membership (well,
+	// apart from the primary key columns), no AUTO_INC column and no FOREIGN
+	// KEY, there is no obvious setting to tweak (based on the recommendations of
+	// the third link above).
+	//
+	// However, in label_membership, updated_at defaults to the current timestamp
+	// both on INSERT and when UPDATEd, so it does not need to be provided.
+
+	runInsertBatch := func(batch [][2]uint) error {
+		sql := `INSERT INTO label_membership (label_id, host_id) VALUES `
+		sql += strings.Repeat(`(?, ?),`, len(batch))
+		sql = strings.TrimSuffix(sql, ",")
+		sql += ` ON DUPLICATE KEY UPDATE updated_at = VALUES(updated_at)`
+
+		vals := make([]interface{}, 0, len(batch)*2)
+		for _, tup := range batch {
+			vals = append(vals, tup[0], tup[1])
+		}
+		return ds.AdhocRetryTx(ctx, func(tx sqlx.ExtContext) error {
+			_, err := tx.ExecContext(ctx, sql, vals...)
+			return errors.Wrap(err, "insert into label_membership")
+		})
+	}
+
+	runDeleteBatch := func(batch [][2]uint) error {
+		rest := strings.Repeat(`UNION ALL SELECT ?, ? `, len(batch)-1)
+		sql := fmt.Sprintf(`
+    DELETE
+      lm
+    FROM
+      label_membership lm
+    JOIN
+      (SELECT ? label_id, ? host_id %s) del_list
+    ON
+      lm.label_id = del_list.label_id AND
+      lm.host_id = del_list.host_id`, rest)
+
+		vals := make([]interface{}, 0, len(batch)*2)
+		for _, tup := range batch {
+			vals = append(vals, tup[0], tup[1])
+		}
+		return ds.AdhocRetryTx(ctx, func(tx sqlx.ExtContext) error {
+			_, err := tx.ExecContext(ctx, sql, vals...)
+			return errors.Wrap(err, "delete from label_membership")
+		})
+	}
```

Comment:
Yeah I agree, if we like this async approach and go forward with it, `runInsertBatch` and `runDeleteBatch` can be moved to the datastore (with a more descriptive name). I'll leave them like that for now until we load-test that in case we try a different approach.